docs/src/index.md

```diff
@@ -0,0 +1,347 @@
+```julia
+function frule(::typeof(foo), args..., Δself, Δargs...; kwargs...)
+    ...
+    return y, ∂y
+end
+```
+
+
+The input to the pushforward is often called the _perturbation_.
+If the function is `y = f(x)` often the pushforward will be written `ẏ = last(frule(f, x, ṡelf, ẋ))`.
+`ẏ` is commonly used to represent the perturbation for `y`.
+
+!!! note
+
+    In the `frule`/pushforward,
+    there is one `Δarg` per `arg` to the original function.
+    The `Δargs` are similar in type/structure to the corresponding inputs `args` (`Δself` is explained below).
+    The `∂y` are similar in type/structure to the original function's output `Y`.
+    In particular if that function returned a tuple then `∂y` will be a tuple of the same size.
+
+### Self derivative `Δself`, `∂self`, `s̄elf`, `ṡelf` etc.
+
+!!! terminology "Δself, ∂self, s̄elf, ṡelf"
+    It is the derivatives with respect to the internal fields of the function.
+    To the best of our knowledge there is no standard terminology for this.
+    Other good names might be `Δinternal`/`∂internal`.
+
+From the mathematical perspective, one may have been wondering what all this `Δself`, `∂self` is.
+Given that a function with two inputs, say `f(a, b)`, only has two partial derivatives:
+``\dfrac{∂f}{∂a}``, ``\dfrac{∂f}{∂b}``.
+Why then does a `pushforward` take in this extra `Δself`, and why does a `pullback` return this extra `∂self`?
+
+The reason is that in Julia the function `f` may itself have internal fields.
+For example a closure has the fields it closes over; a callable object (i.e. a functor) like a `Flux.Dense` has the fields of that object.
+
+**Thus every function is treated as having the extra implicit argument `self`, which captures those fields.**
+So every `pushforward` takes in an extra argument, which is ignored unless the original function has fields.
+It is common to write `function foo_pushforward(_, Δargs...)` in the case when `foo` does not have fields.
+Similarly every `pullback` returns an extra `∂self`, which for things without fields is the constant `NO_FIELDS`, indicating there are no fields within the function itself.
+
+
+### Pushforward / Pullback summary
+
+- **Pullback**
+   - returned by `rrule`
+   - takes output space wobbles, gives input space wiggles
+   - 1 argument per original function return
+   - 1 return per original function argument + 1 for the function itself
+
+- **Pushforward:**
+    - part of `frule`
+    - takes input space wiggles, gives output space wobbles
+    - 1 argument per original function argument + 1 for the function itself
+    - 1 return per original function return
+
+
+### Pullback/Pushforward and Directional Derivative/Gradient
+
+The most trivial use of the `pushforward` from within `frule` is to calculate the directional derivative:
+
+If we would like to know the the directional derivative of `f` for an input change of `(1.5, 0.4, -1)`
+
+```julia
+direction = (1.5, 0.4, -1) # (ȧ, ḃ, ċ)
```

Comment:
Not sure how useful this would be, but as a non-researcher. Having full working examples would be nice to have so I could reproduce and play with them to get a better headspace of what's going on.

If you think it's worth while, logging a future issue is fine for this.

Reply:
I think all these suggestions are good - please open an issue :)

Reply:
https://github.com/JuliaDiff/ChainRulesCore.jl/issues/116